### PR TITLE
feat(sdk): add continueOn support for tasks. Fixes #6749, fixes #7330

### LINKED ISSUE #6749: [feature] continueOn functionality available in Argo but not in KFP
### Feature Area
/area sdk
### What feature would you like to see?
I would like to see the feature described in https://github.com/kubeflow/pipelines/issues/2372 
let me copy/paste my comments from there.
### What is the use case or pain point?
I would be happy to have the same functionality. I've even tested that already by manual change of workflow.yaml:
```
    dag:
      tasks:
      - name: catch-fail
        template: catch-fail
        continueOn:
          failed: true
```
so I might try to propose the implementation looking into how dependencies are handled in the kfp code.
Scenario: we have a pipeline containing a few steps, but some are "optional" which means if they fail it's acceptable for us.
Sure that the best would be then to have pipeline finished with the status "Warning" instead of "Success" but still such an functionality is fine. 

### Is there a workaround currently?
We've implemented an workaround by using dsl.Condition but it's ugly and makes the stuff less readable.

--- 

Love this idea? Give it a 👍. We prioritize fulfilling features with the most 👍.

### LINKED ISSUE #7330: [feature] allow jobs to fail
### Feature Area

/area backend

### What feature would you like to see?

_**Allow failure of jobs**_ - If an operation fails, do not fail the pipeline. Allow the pipeline to continue to the next stage, and there it may fail if that does not have the pre-requisites.

### What is the use case or pain point?

In the machine learning pipelines, it is fairly common to run multiple models, or possibly different configurations of same model, and this possibly runs on a subset of training data. After these are trained, usually they are compared using some metric, the best model is chosen, and that is run on the entire training data to have the final trained model.

If someone uses [`kfp.dsl.ParallelFor`](https://kubeflow-pipelines.readthedocs.io/en/latest/source/kfp.dsl.html#kfp.dsl.ParallelFor) to run the different models, failure in one of them causes the entire pipeline to fail and possibly successful training of other models are lost. But if the next stage, the one to compare using metric supports comparison of the available (i.e. successful) models, the pipeline failure costs the time to train those models, as one have to restart. If we support the requested feature, the failed operations will display an warning (may be ⚠️), and will go on to final training step. Then depending on whether that supports comparison of subset of all models, it will proceed as if the failed models were not there. If not, it'll fail there.

Very similar functionality in available in few CI tools. For example, Gitlab CI has [`allow_failure`](https://docs.gitlab.com/ee/ci/yaml/#allow_failure), Travis CI has [`allow_failures`](https://docs.travis-ci.com/user/customizing-the-build/#jobs-that-are-allowed-to-fail), etc.

### Is there a workaround currently?

It is possible to do very broad top level exception handling to suppress failures. However, in this way the fact that it failed is hidden in the logs and not displayed in the pipeline dashboard. In scheduled pipelines where no one really go through the logs of all "successful" pipelines, these failures will go unnoticed.

---

Love this idea? Give it a 👍. We prioritize fulfilling features with the most 👍.

### PR DESCRIPTION
**Pull request by marrrcin**
**Description of your changes:**
Add support for Agro's continueOn capability to Kubeflow Pipelines. This allows the pipelines to continue execution even though some of the jobs fail, which is really useful, especially when pipelines extensively use `ParallelFor` abstraction and the branches of the `ParallelFor` are independent.

The pipeline with this feature enabled looks like this:
Screenshot 2022-03-03 at 12 12 13
Skipped steps have status: `This step is in Omitted state with this message: omitted: depends condition not met` and the overall pipeline status is Failed, which is expected.

**Implementation details:**
The compiler adds Argo's [continueOn field](https://argoproj.github.io/argo-workflows/fields/#continueon) to the task  if the `ContainerOp` has pod annotation `pipelines.kubeflow.org/allow_failure: true` 

This is for KFP DSL v1, but I can also do the same PR for v2 from master.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -831,7 +837,10 @@ def _create_pipeline_workflow(self,
 
         # set ttl after workflow finishes
         if pipeline_conf.ttl_seconds_after_finished >= 0:
-            workflow['spec']['ttlStrategy'] = {'secondsAfterCompletion': pipeline_conf.ttl_seconds_after_finished}
+            workflow['spec']['ttlStrategy'] = {
```
Comment (marrrcin) [inline sdk/python/kfp/compiler/compiler.py:840]: yapf formatter result

File: sdk/python/kfp/compiler/compiler.py [hunk 2]
```diff
@@ -744,8 +751,7 @@ def _create_dag_templates(self,
         for op in pipeline.ops.values():
             if hasattr(op, 'importer_spec'):
                 raise ValueError(
-                    'dsl.importer is not supported with v1 compiler.'
-                )
+                    'dsl.importer is not supported with v1 compiler.')
```
Comment (marrrcin) [inline sdk/python/kfp/compiler/compiler.py:754]: yapf formatter result

File: sdk/python/tests/compiler/compiler_tests.py [hunk 3]
```diff
@@ -981,8 +1021,9 @@ def some_pipeline():
             dsl.get_pipeline_conf().set_ttl_seconds_after_finished(86400)
 
         workflow_dict = kfp.compiler.Compiler()._compile(some_pipeline)
-        self.assertEqual(workflow_dict['spec']['ttlStrategy']['secondsAfterCompletion'],
-                         86400)
+        self.assertEqual(
```
Comment (marrrcin) [inline sdk/python/tests/compiler/compiler_tests.py:1024]: yapf formatter result

**Comment by marrrcin** [comment]
@chensun  / @ji-yaqi I would really appreciate your input on this PR as our team will greatly benefit from this feature in our pipelines.

**Comment by chensun** [comment]
> This allows the pipelines to continue execution even though some of the jobs fail, which is really useful

I think this is rather subjective. On Vertex Pipelines, the current behavior is what you described--pipeline continue execution when independent components failed. And we actually heard some customer complains on this. It could mean waste of computing resource some time. So I don't think we want to change the default behavior to make it continue on failure. 
That being said, I understand that your proposal could be desirable in certain scenarios. So how about make it an user option instead. I think this could probably be done in `PipelineConf`, here's an example usage:
https://github.com/kubeflow/pipelines/blob/bfffa83fbadd2b25efbeaefa022a95bb02340faa/samples/test/utils/kfp/samples/test/utils.py#L298-L305

**Comment by marrrcin** [comment]
And how does this PR change the default behaviour? It's a completely configurable, opt-in feature that is only enabled if you explicitly add the annotation.

**Comment by misteliy** [comment]
I fully agree with @marrrcin here - could you elaborate your concerns?

**Comment by chensun** [comment]
Sorry about the late reply. 

> And how does this PR change the default behaviour? It's a completely configurable, opt-in feature that is only enabled if you explicitly add the annotation.

@marrrcin You're right. I misread. This doesn't change the default behavior.


> I fully agree with @marrrcin here - could you elaborate your concerns? @chensun @ji-yaqi

@misteliy @jaklan My concern is this PR makes pod annotation as a user interface to configure runtime behavior. We don't have such a pattern, and we likely don't want to go this route as it prevents cross-platform compatibility.
That's why I was suggesting looking into `PipelineConf`.
